Sprint 16

nataliyaTruf opened the pull request:
Completed tasks for sprint 16:

- Refactored the Category List screen to MVVM architecture with bindings via closures.
- Implemented onboarding screens using UIPageViewController
- Refactored CreateTrackerViewController, ScheduleViewController, and TrackersViewController to MVVM architecture using Combine (had to sacrifice code style uniformity to try different approaches to implementing the MVVM pattern)

nataliyaTruf committed:
Implement onboarding flow using UIPageViewController in Tracker App

- Implemented OnboardingViewController with UIPageViewController using the scroll transition style for smooth horizontal paging
- Added functionality to skip the onboarding process. This includes saving the user's preference to skip future onboarding sessions using UserDefaults
- Setup DataSource and Delegate protocols to manage page transitions and define actions for moving between pages
- Ensured that the app directly loads the main interface if the user has completed or skipped onboarding previously

nataliyaTruf committed:
Refactor: Standardize UI Components and Update Controller Implementations for Enhanced Consistency and Maintainability

This commit optimizes the UI component setup across the application and updates various controllers to align with the new UI standards. Key improvements include:

UITableView Extension: Enhanced to automatically apply consistent horizontal constraints and standard style settings. This centralization reduces repetitive boilerplate code across multiple view controllers.
CustomButton and CustomTitleLabel: These components have embedded layout constraints for consistent positioning and sizing, ensuring uniformity across all views and reducing setup overhead
ConfigurableTableViewCell: Developed to support multiple accessory types (none, disclosureIndicator, checkmark, switchControl) and dynamically show or hide elements based on configuration needs, thus providing flexibility in UI management
Controller Adjustments: Updated controllers to utilize the standardized UI components effectively. This includes modifying how tables are set up and interacted with, ensuring that all tables uniformly handle user interactions and display data
Streamlined Integration: Removed redundant views and code, ensuring that all UI elements conform to a unified design language. This enhances code readability and maintainability, making it easier to manage and update in the future
CreateTrackerViewController Update: Replaced the CustomOptionView with UITableView and ConfigurableTableViewCell for category and schedule options. This change utilizes the standardized ConfigurableTableViewCell to handle accessory types and additional text, ensuring a more consistent and flexible UI
These enhancements contribute to a more maintainable, consistent, and scalable codebase, improving the overall development experience and application performance

nataliyaTruf committed:
Implement Category Management

- Created CategoryListViewController and AddCategoryViewController for managing tracker categories
- Integrated category creation and selection functionality within the app
- Ensured category persistence using Core Data, enhancing data integrity and user experience
- Refactored data flow to include category tracking, enabling better organization and retrieval of trackers
- Improved overall user interaction with added category features, making the app more intuitive and organized

nataliyaTruf committed:
Update logic for displaying irregular events in TrackersViewController

nataliyaTruf committed:
Refactor CategoryListViewController to MVVM architecture

- Create CategoryListViewModel to handle data and business logic
- Move category-related data and methods from the view controller to the view model
- Update CategoryListViewController to bind with the view model
- Ensure UI updates and category selection handling through closures
- Maintain proper separation of concerns to adhere to MVVM principles

nataliyaTruf committed:
Refactor AddCategoryViewController to MVVM architecture

- Created AddCategoryViewModel to handle category addition logic
- Moved data and business logic from AddCategoryViewController to AddCategoryViewModel
- Updated AddCategoryViewController to bind with AddCategoryViewModel
- Ensured UI updates through closures for category addition and button state changes
- Maintained separation of concerns by keeping UI logic in the view controller and data logic in the view model

nataliyaTruf committed:
Refactor: Centralize font management and create CustomTextField

nataliyaTruf committed:
Refactor TrackersViewController to MVVM architecture using Combine

Created TrackersViewModel to handle tracker-related data and business logic
Moved data fetching, filtering, and manipulation logic from TrackersViewController to TrackersViewModel
Updated TrackersViewController to bind with TrackersViewModel using Combine
Ensured UI updates through Combine publishers for categories, search state, and current date
Maintained separation of concerns by keeping UI logic in the view controller and data logic in the view model

nataliyaTruf committed:
Refactor CreateTrackerViewController to MVVM architecture using Combine

- Moved logic from CreateTrackerViewController to CreateTrackerViewModel
- Updated UI components to bind with ViewModel using Combine
- Fixed logic for displaying irregular events in TrackersViewController

nataliyaTruf committed:
Refactor ScheduleViewController to MVVM architecture using Combine

- Implement ScheduleViewModel to manage state and logic
- Use Combine framework to handle state changes and bindings

helecta commented on Tracker/Components/ConfigurableTableViewCell.swift:
```diff
@@ -0,0 +1,172 @@
+            checkmarkIcon.isHidden = false
+            selectionStyle = .blue
+        case .switchControl(let isOn):
+                    accessorySwitch.isHidden = false
```
Не забывай следить за отступами, общепринятые отступы помогают читать код быстрее. В общем случае поможет выделить весь файл cmd+A и зажать control+i.

helecta commented on Tracker/Components/CustomTitleLabel.swift:
```diff
@@ -0,0 +1,58 @@
+            }
+        }
+        
+        private func setupConstraintsForStackView() {
```
Не забывай следить за отступами

helecta commented on Tracker/Screens/AddCategory/AddCategoryViewModel.swift:
```diff
@@ -0,0 +1,32 @@
+    
+    // MARK: - Closures
+    
+    var onCategoryAdded: ((String) -> Void)?
```
Не забывай организовывать код следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы
- IBAction

Подробнее почитать можно тут: [Организация кода](https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization)

helecta commented on Tracker/Screens/AddCategory/AddCategoryViewController.swift:
```diff
@@ -0,0 +1,110 @@
+final class AddCategoryViewController: UIViewController {
+    // MARK: - Properties
+    
+    var onCategoryAdded: ((String) -> Void)?
```
Это излишний колбек. 
Сейчас ты в нем сохраняешь новую категорию, но только она у тебя уже сохранена в AddCategoryViewModel в addCategory, и это правильно, за сохранение отвечает AddCategoryViewModel, а не CategoryListViewModel.
Возможно ты хочешь его использовать для обновления списка категорий? Но тогда не стоит повторно сохранять, достаточно вызвать обновление списка категорий:
```
        addCategoryVC.onCategoryAdded = { [weak self] newCategoryName in
            self?.viewModel.loadCategories()
        }
```
а этот метод 
```
    func addCategory(name: String) {
        categoryStore.createCategory(title: name)
        loadCategories()
    }
```
можно удалить

helecta commented on Tracker/Screens/CategoryList/CategoryListViewController.swift:
```diff
@@ -0,0 +1,131 @@
+    }
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        viewModel.selectCategory(at: indexPath.row)
```
у тебя у ячейки есть перехватка жеста onCellTapped, поэтому в этом месте возникает проблема, что при нажатии ничего не происходит, т.к нажатие перехватывается и этот метод уже не вызывается. (жесты которые ты добавляешь идут более приоритетными). 
Можно сделать так, что если не задается onCellTapped, то не добавлять UITapGestureRecognizer

helecta commented on Tracker/Screens/CategoryList/CategoryListViewModel.swift:
```diff
@@ -0,0 +1,52 @@
+        categories = categoryStore.getAllCategoriesWithTrackers()
+    }
+    
+    func addCategory(name: String) {
```
вот этого метода не должно быть, эту задачу должна выполнять AddCategoryViewModel

helecta commented on Tracker/Screens/OnboardingViewController.swift:
```diff
@@ -0,0 +1,167 @@
+    // MARK: - Setup Methods
+    
+    private func createOnboardingViewController(forPage page: Int) -> UIViewController {
+        let pageContentViewController = UIViewController()
```
Лучше было бы сделать отдельный контроллер: OnboardingPageViewController и описать в нем один раз верстку и настройку элементов, а здесь создавать его с разной настройкой.
Вот хороший пример реализации онбординга с енамами: https://github.com/cjlarsen/Tutorials/tree/master/PageControllerTutorial

helecta commented on Tracker/Screens/OnboardingViewController.swift:
```diff
@@ -0,0 +1,167 @@
+    
+    lazy var pages: [UIViewController] = {
+        var pages = [UIViewController]()
+        pages.append(createOnboardingViewController(forPage: 1))
```
Надо запомнить правило: вью контроллеры не надо хранить ссылками в других вью-контроллерах, а создавать их по мере необходимости, то есть при навигации.
Например, когда ты настраиваешь UIPageViewController - используешь setViewControllers -в этот момент создать первый, а когда будешь навигировать каруселью ко второму (через методы делегата), создавать второй, и так далее.

helecta commented on Tracker/Screens/Trackers/TrackersViewController.swift:
```diff
@@ -6,21 +6,26 @@
-import UIKit
+/**
+ По заданию CategoryListViewController переписан на архитектуру MVVM с байндингами через замыкания, но после согласования с наставником, я решила использовать Combine для других контроллеров, чтобы попробовать разные подходы к реализации паттерна MVVM.
+ Таким образом, пришлось пожертвовать однородностью стиля кода ради учебных целей.
```
👍

helecta commented on Tracker/Screens/Trackers/TrackersViewController.swift:
```diff
@@ -165,15 +196,24 @@ final class TrackersViewController: UIViewController {
+    private func toggleTrackerCompleted(trackerId: UUID, at indexPath: IndexPath) {
+        viewModel.toggleTrackerCompleted(trackerId: trackerId)
+        UIView.performWithoutAnimation {
+            self.trackersCollectionView.reloadItems(at: [indexPath])
```
Строгая ссылка на self. Лучше в захвате замыкания сделать слабую

helecta commented on Tracker/Screens/Trackers/TrackersViewController.swift:
```diff
@@ -308,106 +345,42 @@ extension TrackersViewController: UICollectionViewDelegateFlowLayout {
-        trackersCollectionView.reloadData()
+    func trackerRecordStoreDidChangeContent(records: [TrackerRecord]) {
+        DispatchQueue.main.async {
+            self.viewModel.loadCompletedTrackers()
```
В замыкание для избежания retain cycl'ов, необходимо передавать слабую ссылку на `self`

helecta commented on Tracker/Screens/Trackers/TrackersViewController.swift:
```diff
@@ -308,106 +345,42 @@ extension TrackersViewController: UICollectionViewDelegateFlowLayout {
+    func trackerCategoryStoreDidChangeContent(_ store: TrackerCategoryStore) {
         DispatchQueue.main.async {
-            self.loadCompletedTrackers()
+            self.viewModel.loadCategories()
```
В замыкание для избежания retain cycl'ов, необходимо передавать слабую ссылку на `self`

nataliyaTruf committed:
Update .gitignore to exclude Xcode scheme files

nataliyaTruf committed:
Refactor category handling: streamline view models

nataliyaTruf committed:
Centralize tap handling for ConfigurableTableViewCell in tableView(_:didSelectRowAt:)
Removing the need for onCellTapped property for cleaner code

nataliyaTruf committed:
Fix retain cycles in TrackersViewController

nataliyaTruf committed:
Refactor Onboarding Flow with OnboardingPage Enum and Dynamic Page Loading

- Introduce OnboardingPage enum for onboarding page data
- Use OnboardingPageViewController for individual onboarding pages
- Optimize OnboardingViewController with dynamic page loading
- Update UIPageControl and navigation methods for better performance

nataliyaTruf committed:
Refactor: Organize code for better readability

nataliyaTruf committed:
Refactor: Extract EmptyStateView component for code optimization and reuse

nataliyaTruf committed:
Implement EmptyStateView in CategoryListViewController

Refactor to MVVM: added ViewState enum, binding methods for UI updates

nataliyaTruf committed:
Optimize EmptyStateView handling in TrackersViewController

- Refactor to MVVM: Added ViewState enum, improved binding methods for UI updates, and optimized search functionality to handle empty states correctly

nataliyaTruf committed:
Enhance schedule display in CreateTrackerVC to show specific texts for selected days

nataliyaTruf committed:
Enhance schedule selection functionality to retain selected days on reopening
